File: Pod/Classes/HSM/CPMessageOperation.m
//
//  CPSendMessageOperation.m
//  GCD的其他用法
//
//  Created by cqh on 15/1/27.
//  Copyright (c) 2015年 Cnepay. All rights reserved.
//

#import "CPMessageOperation.h"
NSString* const MessageMapping [] = {
    [MessageTypeEventInit] = @"MessageTypeEventInit",
    [MessageTypeEventRetryRequest] = @"MessageTypeEventRetryRequest",
    [MessageTypeEventVerifyPass] = @"MessageTypeEventVerifyPass",
    [MessageTypeEventVerifyFail] = @"MessageTypeEventVerifyFail",
    [MessageTypeEventDownloadingAvailable] = @"MessageTypeEventDownloadingAvailable",
    [MessageTypeEventProgress] = @"MessageTypeEventProgress",
    [MessageTypeEventTaskDone] = @"MessageTypeEventTaskDone",
    
    [MessageTypeEventDownloadException] = @"MessageTypeEventDownloadException",
    [MessageTypeEventNetworkConnectionChange] = @"MessageTypeEventNetworkConnectionChange",
    
    [MessageTypeActionStart] = @"MessageTypeActionStart",
    [MessageTypeActionPaused] = @"MessageTypeActionPaused",
    [MessageTypeActionDelete] = @"MessageTypeActionDelete",
    [MessageTypeActionQuit] = @"MessageTypeActionQuit",
};
@interface CPMessageOperation ()
/*
 * 利用无名扩展 让readOnly属性,在内部可以变成可读可写属性。
 */
@property (assign, nonatomic, getter=isExecuting) BOOL executing;
@property (assign, nonatomic, getter=isFinished) BOOL finished;
@end

@implementation CPMessageOperation
@synthesize executing = _executing;
@synthesize finished = _finished;

+ (instancetype)sendMessageOperationWithTask:(CPMessage *)message {
    
    return [[self alloc] initWithTask:message];
}


- (instancetype)initWithTask:(CPMessage *)message {
    self = [super init];
    if (self) {
        self.message = message;
    }
    return self;
}

- (void)setFinished:(BOOL)finished
{
    [self willChangeValueForKey:@"isFinished"];
    _finished = finished;
    [self didChangeValueForKey:@"isFinished"];
}

- (void)setExecuting:(BOOL)executing {
    [self willChangeValueForKey:@"isExecuting"];
    _executing = executing;
    [self didChangeValueForKey:@"isExecuting"];
}


- (void)start
{
    @synchronized(self) {
        if(self.isCancelled)
        {
            self.finished = YES;
            return;

        } else if ([self isReady]){
            self.executing = YES;
            [self performSelector:@selector(operationDidStart) onThread:[self runloopThread] withObject:nil waitUntilDone:NO modes:@[NSRunLoopCommonModes]];
        }
    }
}

- (void) operationDidStart
{
    if (self.delayTime != 0.0) {
        [self performSelector:@selector(operationDidStart) withObject:nil afterDelay:self.delayTime inModes:@[NSRunLoopCommonModes]];
        self.delayTime = 0.0;
    } else {
        if ([self.delegate respondsToSelector:@selector(sendMessageOperationDidStart:message:)]) {
            [self.delegate sendMessageOperationDidStart:self message:self.message];
            self.finished = YES;
        }
    }
}

//- (void)main {
//    @autoreleasepool {
//      //发送消息
//        if(self.isCancelled) return;
//        if ([self.delegate respondsToSelector:@selector(sendMessageOperation:didFinishSendMessage:)]) {
//            [self.delegate sendMessageOperation:self didFinishSendMessage:self.message];
//        }
//    }
//}

- (void)dealloc {
//    NSLog(@"--%@ 挂掉了",self.class);
}
@end


@implementation CPMessage

- (instancetype)initWithType:(MessageType)type obj:(id)object {
    return [self initWithType:type subType:0 code:0 obj:object];
}
+ (instancetype)messageWithType:(MessageType)type obj:(id)object {
    return  [[self alloc] initWithType:type obj:object];
}
+ (instancetype)messageWithType:(MessageType)type
{
    return [[self alloc] initWithType:type subType:0 code:0 obj:nil];
}
- (instancetype)initWithType:(MessageType)type
{
    return [self initWithType:type subType:0 code:0 obj:nil];
}


+ (instancetype)messageWithType:(MessageType)type subType:(int)subType code:(int)code obj:(id)object
{
    return [[self alloc] initWithType:type subType:subType code:code obj:object];
}

- (instancetype)initWithType:(MessageType)type subType:(int)subType code:(int)code obj:(id)object
{
    if (self == [super init]) {
        _type = type;
        _subType = subType;
        _code = code;
        _obj = object;
    }
    return self;
}




- (NSString *)description {
    return [NSString stringWithFormat:@"type:%zd subType:%d code:%d id:%@",_type,_subType,_code,_obj];
}


@end

File: Pod/Classes/HSM/CPMessageOperation.swift
//
//  CPMessageOperation.swift
//  StateMachine-Swift
//
//  Created by Cnepay on 15/10/18.
//  Copyright © 2015年 Cnepay. All rights reserved.
//

import UIKit


class CPMessage {
    
    var type :MESSAGETYPE?// = MESSAGETYPE.ACTION_CHOOSEDEVICE
    var subType:Int?
    var code = 0
    
    var obj:AnyObject?
}

//func  == (left:CPMessage,right:CPMessage) -> Bool {
//    
//    return (left.type == right.type);
//}
//
//func !=(left:CPMessage,right:CPMessage) -> Bool {
//    
//    return !(left == right);
//}


protocol CPMessageOperationDelegate {
    func sendMessageOperationDidStart(operation:CPMessageOperation,message:CPMessage) ->()
}

class CPMessageOperation: NSOperation {
    var message:CPMessage;
    var runLoopThread:NSThread!;
    var delegate:CPMessageOperationDelegate?
    
    
    private var _executing: Bool = false
    override var executing: Bool {
        get {
            return _executing
        }
        set {
            if _executing != newValue {
                willChangeValueForKey("isExecuting")
                _executing = newValue
                didChangeValueForKey("isExecuting")
            }
        }
    }
    
    private var _finished: Bool = false;
    override var finished: Bool {
        get {
            return _finished
        }
        set {
            if _finished != newValue {
                willChangeValueForKey("isFinished")
                _finished = newValue
                didChangeValueForKey("isFinished")
            }
        }
    }

    
    init(message:CPMessage) {
        self.message = message;
        super.init()
    }
    
//    static var sendDeviceCommandThread:NSThread = {
//        let thread =  NSThread(target: CPMessageOperation.self, selector: "sendDeviceCommandThreadEntryPoint:", object: nil);
//        thread.start();
//        return thread;
//    }();
    
    
//    static func sendDeviceCommandThreadEntryPoint(thread:NSThread) ->() {
//        let currentThread = NSThread.currentThread();
//        currentThread.name = "SendDeviceCommand"
//        let currentRunLoop = NSRunLoop.currentRunLoop();
//        currentRunLoop.addPort(NSMachPort(), forMode: NSDefaultRunLoopMode);
//        currentRunLoop.run();
//    }
    
    override func start() {
        
        if self.cancelled {
            self.finished = true
            return
            
        }else if self.ready {
            self.executing = true
            self.performSelector("operationDidStart", onThread: runLoopThread, withObject: nil, waitUntilDone:false)
        }
    }
    
    func operationDidStart() -> (){
        self.delegate?.sendMessageOperationDidStart(self, message: self.message);
        self.finished = true;
    }
}


